### PR TITLE
add contexts to the flexible metadata model.

### PR DESCRIPTION
**Pull request by orangewolf**
they can be assigned on admin sets and then apply the contexts to the works made with that admin set selected

### CODE REVIEW

File: lib/generators/hyrax/templates/db/migrate/20240606205216_add_contexts_to_hyrax_flexible_schemas.rb.erb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class AddContextsToHyraxFlexibleSchemas < ActiveRecord::Migration<%= migration_version %>
+  def change
```
Comment (laritakr) [inline lib/generators/hyrax/templates/db/migrate/20240606205216_add_contexts_to_hyrax_flexible_schemas.rb.erb:2]: Should this migration have a conditional check if it already exists?

File: app/models/hyrax/flexible_schema.rb [hunk 2]
```diff
@@ -43,6 +50,10 @@ def schema_version
     profile['m3_version']
   end
 
+  def context_select
+    contexts.map { |k, v| [v&.[]('display_label'), k] }
```
Comment (laritakr) [inline app/models/hyrax/flexible_schema.rb:54]:
```suggestion
// contexts&.map { |k, v| [v&.[]('display_label'), k] }
```

File: config/metadata_profiles/m3_profile.yaml [hunk 3]
```diff
@@ -18,6 +18,8 @@ classes:
 contexts:
   flexible_context:
     display_label: Flexible Metadata Example
+  special_context:
```
Comment (laritakr) [inline config/metadata_profiles/m3_profile.yaml:21]: Corresponding changes to this profile should probably also be made in the .koppie (and possibly .dassie?) versions, as well as the spec version.

File: app/services/hyrax/m3_schema_loader.rb [hunk 4]
```diff
@@ -21,11 +21,19 @@ def view_definitions_for(schema:, version: 1)
     ##
     # @param [#to_s] schema_name
     # @return [Enumerable<AttributeDefinition]
-    def definitions(schema_name, version)
+    def definitions(schema_name, version, contexts = nil)
       schema = Hyrax::FlexibleSchema.find_by(id: version) || Hyrax::FlexibleSchema.create_default_schema
       schema.attributes_for(schema_name).map do |name, config|
+        # We might be able to consolidate these conditions, but they have been kept seperate to make it easier to reason about
```
Comment (laritakr) [inline app/services/hyrax/m3_schema_loader.rb:27]:
```suggestion
# We might be able to consolidate these conditions, but they have been kept separate to make it easier to reason about
```

**Review by laritakr — CHANGES_REQUESTED** [review]
A few comments... 
- the M3 profiles in specs, .koppie, and .dassie do not match the changes in the config profile. These need to be updated
- no specs exist for the code changes here
- the migration needs to also be added to koppie & dassie
- Admin set show page should display the metadata context
- I was unable to save a work with the dimensions term

This is the error I got trying to save the work using the context term. I'm not certain of the cause, [but it originates on this line](https://github.com/samvera/valkyrie/blob/main/lib/valkyrie/persistence/postgres/persister.rb#L34C1-L34C55) when it calls `definitions` in the m3_schema_loader.

Screenshot 2024-10-08 at 5 40 09 PM

File: app/forms/concerns/hyrax/flexible_form_behavior.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    extend ActiveSupport::Concern
+
+    included do
+      include Hyrax::BasedNearFieldBehavior
+      property :contexts
+    end
+
+    # OVERRIDE disposable 0.6.3 to make schema dynamic
+    def schema
+      if Hyrax.config.flexible?
```
[the remark this answers is not in the document]
Comment (orangewolf) [reply app/forms/concerns/hyrax/flexible_form_behavior.rb:13]: correct, I'll take them out

File: lib/generators/hyrax/templates/db/migrate/20240606205216_add_contexts_to_hyrax_flexible_schemas.rb.erb (hunk 1 above)
Comment (orangewolf) [reply lib/generators/hyrax/templates/db/migrate/20240606205216_add_contexts_to_hyrax_flexible_schemas.rb.erb:2]: yep

**Pull request merged**